safekeeper/src/timeline.rs

```diff
@@ -589,6 +592,80 @@ impl Timeline {
+    /// mix of WAL from different histories. Term is remembered in the
+    /// BasebackupContext and checked in finish_basebackup.
+    /// Note that in the last segment some WAL higher than flush_lsn set
+    /// here might be streamed; that's fine as long as terms doesn't change.
+    ///
+    /// In the future we might send only up to commit_lsn to get some valid
+    /// state which later will be recovered by compute, in this case term check
+    /// is not needed.
+    ///
+    /// When returned BasebackupContext is dropped WAL hold is removed.
+    pub async fn start_basebackup<W: AsyncWrite + Unpin + Send>(
+        self: &Arc<Self>,
+        ar: &mut tokio_tar::Builder<W>,
+    ) -> Result<BasebackupContext> {
+        let shared_state = self.read_shared_state().await;
+
+        let cf_path = self.timeline_dir.join(CONTROL_FILE_NAME);
+        let mut cf = File::open(cf_path).await?;
+        ar.append_file(CONTROL_FILE_NAME, &mut cf).await?;
+
+        // We need to stream since the oldest segment someone (s3 or pageserver)
+        // still needs. This duplicates calc_horizon_lsn logic.
+        let from_lsn = min(
+            shared_state.sk.state.remote_consistent_lsn,
+            shared_state.sk.state.backup_lsn,
+        );
+        let from_segno = from_lsn.segment_number(shared_state.get_wal_seg_size());
+        let term = shared_state.sk.get_term();
+        let last_log_term = shared_state.sk.get_last_log_term();
+        let flush_lsn = shared_state.sk.flush_lsn();
+        let upto_segno =
+            flush_lsn.segment_number(shared_state.sk.state.server.wal_seg_size as usize);
+
+        // TODO: set WAL hold off.
+
+        let bctx = BasebackupContext {
+            from_segno,
+            upto_segno,
+            term,
+            last_log_term,
+            flush_lsn,
+            wal_seg_size: shared_state.get_wal_seg_size(),
+            tli: self.clone(),
+        };
+
+        Ok(bctx)
+    }
+
+    /// Finish basebackup: check that term(s) hasn't changed.
+    ///
+    /// Note that WAL gc hold off is removed in Drop of BasebackupContext to not
+    /// forget this if backup fails mid the way.
+    pub async fn finish_basebackup(self: &Timeline, bctx: &BasebackupContext) -> Result<()> {
+        let shared_state = self.read_shared_state().await;
+        let term = shared_state.sk.get_term();
+        let last_log_term = shared_state.sk.get_last_log_term();
+        // There are some cases to relax this check (e.g. last_log_term might
+        // change, but as long as older history is strictly part of new that's
+        // fine), but there is no need to do it.
+        if bctx.term != term || bctx.last_log_term != last_log_term {
+            bail!("term(s) changed during basebackup: were term={}, last_log_term={}, now term={}, last_log_term={}",
+              bctx.term, bctx.last_log_term, term, last_log_term);
+        }
+        Ok(())
+    }
+
```

Comment:
`start_basebackup` and `finish_basebackup` are needed for WAL fetching which requires WAL to be present on disk. My idea is that all WAL accesses should go through `FullAccessTimeline`, I'd move these functions under `FullAccessTimeline` instead of `Arc<Timeline>`. Otherwise they may fail in the future, when offloading will be implemented.

As an alternative, in my previous PR I statred cleaning up `timeline.rs` and moved everything I could to other files. What do you think, can these functions be moved to other file, perhaps `pull_timeline.rs`?

Reply:
See this comment:
https://github.com/neondatabase/neon/pull/7944#discussion_r1633191647

As for precise location, not sure. Argument for timeline.rs is that these functions need shared state access. Argument for  pull_timeline.rs is that it is about pull_timeline.rs. Either works for me.

Reply:
I think I allowed shared state access from all files several commits ago. It's not very good in terms of blurred responsibilities for maintaining the correct state, but it allowed to separate code much better and to reduce the size of `timeline.rs`, which growed very large. The current state is that all code is allowed to read/write shared state, but consensus-related state should not be modified outside of `safekeeper.rs`.

I'd vote for moving functions to `pull_timeline.rs` or creating a separate `snapshot.rs` file for handling snapshot creation.

Reply:
Moved to pull_timeline.rs. Note though that it required to make get_wal_seg_size to be pub(crate).